File: mpm/python/usrp_mpm/chips/lmk04832.py
#
# Copyright 2019-2021 Ettus Research, a National Instruments Brand
#
# SPDX-License-Identifier: GPL-3.0-or-later
#
# This class also has a C++ counterpart in host/lib/usrp/common/lmk04832.cpp and
# host/lib/include/uhdlib/usrp/common/lmk04832.hpp
# If any changes are needed in this file, consider if making the change in the C++ counterpart is
# also necessary.
"""LMK04832 parent driver class."""

import time

from usrp_mpm.mpmlog import get_logger


class LMK04832:
    """Generic driver class for LMK04832 access.

    This is subclassed for the X4xx-specific use case.
    """

    LMK_CHIP_ID = 6
    LMK_PROD_ID = 0xD163

    VCXO_FREQUENCIES = [122.88e6, 100.00e6]
    LMK_VCO0_RANGE_MIN = 2440e6
    LMK_VCO0_RANGE_MAX = 2580e6
    LMK_VCO1_RANGE_MIN = 2945e6
    LMK_VCO1_RANGE_MAX = 3255e6

    # PLL2 Prescaler is in range from 2, 8
    PLL2_PRESCALER = range(2, 9)

    # fmt: off
    RB_PLL1_LD_LOST_MASK = 0b1000
    RB_PLL1_LD_MASK      = 0b0100
    RB_PLL2_LD_LOST_MASK = 0b0010
    RB_PLL2_LD_MASK      = 0b0001
    # fmt: on

    def __init__(self, regs_iface, parent_log=None):
        """Initialize class."""
        self.log = (
            parent_log.getChild("LMK04832") if parent_log is not None else get_logger("LMK04832")
        )
        self.regs_iface = regs_iface
        assert hasattr(self.regs_iface, "peek8")
        assert hasattr(self.regs_iface, "poke8")
        self.poke8 = regs_iface.poke8
        self.peek8 = regs_iface.peek8
        self.enable_3wire_spi = False

    def pokes8(self, addr_vals):
        """Apply a series of pokes.

        pokes8([(0,1),(0,2)]) is the same as calling poke8(0,1), poke8(0,2).
        """
        for addr, val in addr_vals:
            self.poke8(addr, val)

    def get_chip_id(self):
        """Read back the chip ID."""
        chip_id = self.peek8(0x03)
        self.log.trace("Chip ID Readback: {}".format(chip_id))
        return chip_id

    def get_product_id(self):
        """Read back the product ID."""
        prod_id_0 = self.peek8(0x04)
        prod_id_1 = self.peek8(0x05)
        prod_id = (prod_id_1 << 8) | prod_id_0
        self.log.trace("Product ID Readback: 0x{:X}".format(prod_id))
        return prod_id

    def verify_chip_id(self):
        """Return True if the chip ID and product ID matches what we expect.

        Returns False otherwise.
        """
        chip_id = self.get_chip_id()
        prod_id = self.get_product_id()
        if chip_id != self.LMK_CHIP_ID:
            self.log.error("Wrong Chip ID 0x{:X}".format(chip_id))
            return False
        if prod_id != self.LMK_PROD_ID:
            self.log.error("Wrong Product ID 0x{:X}".format(prod_id))
            return False
        return True

    def enable_4wire_spi(self):
        """Enable 4-wire SPI readback from the CLKin_SEL0 pin."""
        self.poke8(0x148, 0x33)
        self.enable_3wire_spi = False

    def get_status(self):
        """Returns PLL lock status as a dictionary."""
        pll1_status = self.check_plls_locked(pll="PLL1")
        pll2_status = self.check_plls_locked(pll="PLL2")
        return {"PLL1 lock": pll1_status, "PLL2 lock": pll2_status}

    def clear_ldl(self, pll="BOTH"):
        """Clear the LD_LOST flags for both PLLs.

        Set `pll` to "PLL1", "PLL2", or "BOTH" to specify which PLL's LD_LOST
        flags to clear.
        """
        pll = pll.upper()
        assert pll in ("BOTH", "PLL1", "PLL2"), "Invalid PLL specified"
        clear_val1 = 0b10 if pll in ("BOTH", "PLL1") else 0
        clear_val2 = 0b01 if pll in ("BOTH", "PLL2") else 0
        self.poke8(0x182, clear_val1 | clear_val2)
        self.poke8(0x182, 0x00)

    def check_plls_locked(self, pll="BOTH", sticky=False):
        """Return True if the specified PLLs are locked, False otherwise.

        Note: By default, this checks the current status only, it does not check
        the sticky lock-detect bits (i.e., we only check RB_PLL1_LD and/or
        RB_PLL2_LD). If sticky is set to True, this will also check the sticky
        bits, and will clear them afterwards.

        From the datasheet (Section 8.6.2.9.3), register 0x183 bits:
        3: RB_PLL1_LD_LOST. This is set when PLL1 DLD edge falls. Does not set
           if cleared while PLL1 DLD
        2: RB_PLL1_LD. Read back 0: PLL1 DLD is low.  Read back 1: PLL1 DLD is
           high.
        1: RB_PLL2_LD_LOST. This is set when PLL2 DLD edge falls. Does not set
           if cleared while PLL2 DLD
        0: RB_PLL2_LD. Read back 0: PLL2 DLD is low.  Read back 1: PLL2 DLD is
           high.

        The *_LD_LOST bits get cleared by writing to register 0x182 (see
        self.clear_ldl()).
        """
        pll = pll.upper()
        assert pll in ("BOTH", "PLL1", "PLL2"), "Invalid PLL specified"
        result = True
        clear_ldl = False
        pll_lock_status = self.peek8(0x183)
        pll1_ld = bool(pll_lock_status & self.RB_PLL1_LD_MASK)
        pll1_ld_lost = bool(pll_lock_status & self.RB_PLL1_LD_LOST_MASK)
        pll2_ld = bool(pll_lock_status & self.RB_PLL2_LD_MASK)
        pll2_ld_lost = bool(pll_lock_status & self.RB_PLL2_LD_LOST_MASK)

        if pll in ("BOTH", "PLL1"):
            result = result and pll1_ld and (not sticky or not pll1_ld_lost)
            clear_ldl = clear_ldl or (sticky and pll1_ld_lost)
        if pll in ("BOTH", "PLL2"):
            result = result and pll2_ld and (not sticky or not pll2_ld_lost)
            clear_ldl = clear_ldl or (sticky and pll2_ld_lost)
        if not result:
            self.log.debug(
                f"Lock loss reported: RB_PLL1_LD_LOST={int(pll1_ld_lost)},"
                f"RB_PLL1_LD={int(pll1_ld)},RB_PLL2_LD_LOST={int(pll2_ld_lost)},"
                f"RB_PLL2_LD={int(pll2_ld)}"
            )
        if clear_ldl:
            self.clear_ldl(pll)
        return result

    def wait_for_pll_lock(self, pll="BOTH", timeout=2000):
        """Wait for the PLL(s) to lock.

        Returns False if the PLL(s) do not lock before the timeout (in ms)
        """
        self.clear_ldl()
        # Now poll lock status until timeout
        end_time = time.monotonic() + (timeout / 1000)
        while time.monotonic() < end_time:
            time.sleep(0.1)
            if self.check_plls_locked(pll):
                return True
        pll = "PLL1 or PLL2" if pll.upper() == "BOTH" else pll
        self.log.debug("{} not reporting locked after {} ms wait".format(pll, timeout))
        return False

    def soft_reset(self, value=True):
        """Perform soft reset of the LMK04832.

        This is done by setting or unsetting the reset register.
        """
        reset_addr = 0
        if value:  # Reset
            reset_byte = 0x80
        else:  # Clear Reset
            reset_byte = 0x00
        if not self.enable_3wire_spi:
            reset_byte |= 0x10
        self.poke8(reset_addr, reset_byte)

    def pll1_r_divider_sync(self, sync_pin_callback):
        """Run PLL1 R Divider sync.

        Follows the sequence in
        http://www.ti.com/lit/ds/snas688c/snas688c.pdf chapter 8.3.1.1

        Rising edge on sync pin is done by an callback which has to return its
        success state. If the sync callback was successful, returns PLL1 lock
        state as overall success otherwise the method fails.
        """
        # 1) Setup device for synchronizing PLL1 R
        # PLL1R_SYNC_EN    (6) = 1
        # PLL1R_SYNC_SRC (5,4) = Sync pin
        # PLL2R_SYNC_EN    (3) = 0
        self.poke8(0x145, 0x50)

        # Do NOT change clkin0_TYPE and Clkin[0,1]_DEMUX.
        # Both are set in initialization and remain static.

        # 2) Arm PLL1 R divider for synchronization
        self.poke8(0x177, 0x20)
        self.poke8(0x177, 0)
        # 3) wait for the writes to complete by triggering a read
        self.get_chip_id()

        # 4) Send rising edge on SYNC pin
        result = sync_pin_callback()

        # 5) reset 0x145 to safe value (no sync enable set, sync src invalidated)
        self.poke8(0x145, 0)

        # 6) wait for PLL1 to lock
        if result:
            return self.wait_for_pll_lock("PLL1")
        return False

    ## Register bitfield definitions ##

    def pll2_pre_to_reg(self, prescaler, osc_field=0x01, xtal_en=0x0, ref_2x_en=0x0):
        """Return prescaler value.

        From the prescaler value, returns the register value combined with the other
        register fields.
        """
        # valid prescaler values are 2-8, where 8 is represented as 0x00.
        assert prescaler in range(2, 8 + 1)
        reg_val = (
            ((prescaler & 0x07) << 5)
            | ((osc_field & 0x3) << 2)
            | ((xtal_en & 0x1) << 1)
            | ((ref_2x_en & 0x1) << 0)
        )
        self.log.trace(
            "From prescaler value 0d{}, writing register as 0x{:02X}.".format(prescaler, reg_val)
        )
        return reg_val

File: host/lib/usrp/common/lmk04832.cpp
//
// Copyright 2025 Ettus Research, A National Instruments Brand
//
// SPDX-License-Identifier: GPL-3.0-or-later
//

#include <uhd/utils/log.hpp>
#include <uhdlib/usrp/common/lmk04832.hpp>
#include <cassert>
#include <chrono>
#include <thread>

namespace {
constexpr char LOG_ID[]                    = "_LMK04832";
constexpr uint8_t LMK_CHIP_ID              = 6;
constexpr uint16_t LMK_PROD_ID             = 0xD163;
constexpr uint8_t LMK_RB_PLL1_LD_LOST_MASK = 0b1000;
constexpr uint8_t LMK_RB_PLL1_LD_MASK      = 0b0100;
constexpr uint8_t LMK_RB_PLL2_LD_LOST_MASK = 0b0010;
constexpr uint8_t LMK_RB_PLL2_LD_MASK      = 0b0001;
} // namespace

class lmk04832_impl : public lmk04832_iface
{
public:
    lmk04832_impl(write_fn_t&& poke8, read_fn_t&& peek8, std::string log_prefix)
        : _poke8(std::move(poke8))
        , _peek8(std::move(peek8))
        , _enable_3wire_spi(false)
        , _log_id(std::move(log_prefix) + std::string(LOG_ID))
    {
    }

    void pokes8(const std::vector<std::pair<uint16_t, uint8_t>>& addr_val_pair) override
    {
        for (const auto& pair : addr_val_pair) {
            _poke8(pair.first, pair.second);
        }
    }

    uint8_t get_chip_id() override
    {
        uint8_t chip_id = _peek8(0x03);
        UHD_LOG_TRACE(_log_id, "Chip ID Readback: " << int(chip_id));
        return chip_id;
    }

    uint16_t get_product_id() override
    {
        uint8_t prod_id_0 = _peek8(0x04);
        uint8_t prod_id_1 = _peek8(0x05);
        uint16_t prod_id  = (uint16_t(prod_id_1) << 8) | prod_id_0;
        UHD_LOG_TRACE(_log_id, "Product ID Readback: 0x" << std::hex << prod_id);
        return prod_id;
    }

    bool verify_chip_id() override
    {
        uint8_t chip_id  = get_chip_id();
        uint16_t prod_id = get_product_id();
        if (chip_id != LMK_CHIP_ID) {
            UHD_LOG_ERROR(_log_id, "Wrong Chip ID 0x" << std::hex << int(chip_id));
            return false;
        }
        if (prod_id != LMK_PROD_ID) {
            UHD_LOG_ERROR(_log_id, "Wrong Product ID 0x" << std::hex << prod_id);
            return false;
        }
        return true;
    }

    void enable_4wire_spi() override
    {
        _poke8(0x148, 0x33);
        _enable_3wire_spi = false;
    }

    void clear_ldl(lmk04832_pll_sel pll = lmk04832_pll_sel::BOTH) override
    {
        uint8_t clear_val1 =
            (pll == lmk04832_pll_sel::BOTH || pll == lmk04832_pll_sel::PLL1) ? 0b10 : 0;
        uint8_t clear_val2 =
            (pll == lmk04832_pll_sel::BOTH || pll == lmk04832_pll_sel::PLL2) ? 0b01 : 0;
        _poke8(0x182, clear_val1 | clear_val2);
        _poke8(0x182, 0);
    }

    bool check_plls_locked(
        lmk04832_pll_sel pll = lmk04832_pll_sel::BOTH, bool sticky = false) override
    {
        bool result             = true;
        bool clear_ldl_status   = false;
        uint8_t pll_lock_status = _peek8(0x183);
        bool pll1_ld            = pll_lock_status & LMK_RB_PLL1_LD_MASK;
        bool pll1_ld_lost       = pll_lock_status & LMK_RB_PLL1_LD_LOST_MASK;
        bool pll2_ld            = pll_lock_status & LMK_RB_PLL2_LD_MASK;
        bool pll2_ld_lost       = pll_lock_status & LMK_RB_PLL2_LD_LOST_MASK;

        if (pll == lmk04832_pll_sel::BOTH || pll == lmk04832_pll_sel::PLL1) {
            result           = result && pll1_ld && (!sticky || !pll1_ld_lost);
            clear_ldl_status = clear_ldl_status || (sticky && pll1_ld_lost);
        }
        if (pll == lmk04832_pll_sel::BOTH || pll == lmk04832_pll_sel::PLL2) {
            result           = result && pll2_ld && (!sticky || !pll2_ld_lost);
            clear_ldl_status = clear_ldl_status || (sticky && pll2_ld_lost);
        }

        if (!result) {
            UHD_LOG_DEBUG(_log_id,
                "Lock loss reported: (PLL1 LD: "
                    << pll1_ld << ", PLL1 LD Lost: " << pll1_ld_lost << ", PLL2 LD: "
                    << pll2_ld << ", PLL2 LD Lost: " << pll2_ld_lost << ")");
        }

        if (clear_ldl_status) {
            clear_ldl(pll);
        }
        return result;
    }

    bool wait_for_pll_lock(const lmk04832_pll_sel pll = lmk04832_pll_sel::BOTH,
        uint32_t timeout                              = 2000) override
    {
        clear_ldl();

        // Now poll lock status until timeout.
        auto end_time =
            std::chrono::steady_clock::now() + std::chrono::milliseconds(int(timeout));
        while (std::chrono::steady_clock::now() < end_time) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
            if (check_plls_locked(pll)) {
                return true;
            }
        }
        UHD_LOG_DEBUG(
            _log_id, pll << " not reporting locked after " << timeout << " ms wait");
        return false;
    }

    void soft_reset(bool value = true) override
    {
        uint8_t reset_byte = value ? 0x80 : 0x00;
        if (!_enable_3wire_spi) {
            reset_byte |= 0x10;
        }
        _poke8(0, reset_byte);
    }

    void init_pll1_r_divider_sync() override
    {
        // Run PLL1 R Divider sync according to
        // http://www.ti.com/lit/ds/snas688c/snas688c.pdf chapter 8.3.1.1.

        // Rising edge on sync pin is done by an callback which has to return its success
        // state. If the sync callback was successful, returns PLL1 lock state as overall
        // success otherwise the method fails.

        // 1) Setup device for synchronizing PLL1 R.
        // PLL1R_SYNC_EN    (6) = 1
        // PLL1R_SYNC_SRC (5,4) = Sync pin
        // PLL2R_SYNC_EN    (3) = 0
        _poke8(0x145, 0x50);

        // Do NOT change clkin0_TYPE and Clkin[0,1]_DEMUX.
        // Both are set in initialization and remain static.

        // 2) Arm PLL1 R divider for synchronization.
        _poke8(0x177, 0x20);
        _poke8(0x177, 0);
        // 3) Wait for the writes to complete by triggering a read.
        get_chip_id();
        // Caller now needs to send a rising edge on the sync port.
    }

    bool deinit_pll1_r_divider_sync() override
    {
        // 1) reset 0x145 to safe value (no sync enable set, sync src invalidated)
        _poke8(0x145, 0);

        // 2) wait for PLL1 to lock
        return wait_for_pll_lock(lmk04832_pll_sel::PLL1);
    }

    uint8_t pll2_pre_to_reg(uint8_t prescaler,
        uint8_t osc_field = 0x01,
        uint8_t xtal_en   = 0x0,
        uint8_t ref_2x_en = 0x0) override
    {
        assert(prescaler >= 2 && prescaler <= 8);
        uint8_t reg_val = ((prescaler & 0x07) << 5) | ((osc_field & 0x3) << 2)
                          | ((xtal_en & 0x1) << 1) | ((ref_2x_en & 0x1) << 0);
        UHD_LOG_TRACE(_log_id,
            "From prescaler value " << int(prescaler) << ", writing register as 0x"
                                    << std::hex << int(reg_val));
        return reg_val;
    }

private:
    write_fn_t _poke8;
    read_fn_t _peek8;
    bool _enable_3wire_spi;
    std::string _log_id;
};

lmk04832_iface::sptr lmk04832_iface::make(
    write_fn_t&& poke8, read_fn_t&& peek8, std::string log_prefix)
{
    return std::make_shared<lmk04832_impl>(
        std::move(poke8), std::move(peek8), std::move(log_prefix));
}
